### ISSUE TITLE
Does not infer information based on previous bitwise operations when in match statment 

### ISSUE THREAD

**Issue by x4exr** [rust-lang/rust-clippy#12773]
### Summary

Rust does not realize that I performed a bit mask on a number and does not assume the maximum range it can have. doing a & 0b00000001 means the maximum range is 0-1 thus the match should only require 2 arms.

### Lint Name

non-exhaustive patterns

### Reproducer

I tried this code:

```rust
pub fn decode(stream: &mut impl Read, instruction: &mut Instruction) -> Result<(), Error> {
	let mut control_bytes = [0u8; 2];
	match stream.read(&mut control_bytes) {
		Err(error) => return Err(Error::Stream(error)),
		Ok(result) => {
			if result != control_bytes.len() {
				return Err(Error::Supply);
			}
		}
	};
	
	let classification = control_bytes[0] >> 1;
	// Only allow the last bit to be read.
	let destination = match control_bytes[0] & 0b0000000_1 {
		0 => Destination::First,
		1 => Destination::Second,
                // ERROR: Arm expected here
	};
	
	Ok(())
}
```

I saw this happen:

```
non-exhaustive patterns: `2_u8..=u8::MAX` not covered [E0004] pattern `2_u8..=u8::MAX` not covered Note: the matched value is of type `u8` Help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
```

I expected to see this happen:
Successfull compilation

### Version

```text
rustc 1.77.0-nightly (e51e98dde 2023-12-31)
binary: rustc
commit-hash: e51e98dde6a60637b6a71b8105245b629ac3fe77
commit-date: 2023-12-31
host: x86_64-pc-windows-msvc
release: 1.77.0-nightly
LLVM version: 17.0.6
```
